### PR TITLE
feat!: Add refresh tokens

### PR DESCRIPTION
**Pull request by pzl111**
Addresses #88

### Description

Currently we only have accessTokens, which do not expire and poses a security vulnerability as whoever obtains the accessToken can access our services indefinitely.
We implement the access-refresh token workflow. (Refer to additional information below to see how it works).
- Access Tokens: Tokens issued during login and regeneration, short-lived tokens used gain accesss to other services
- Refresh Tokens: Tokens issued ONLY during login, long-lived tokens used to regenerate accessTokens.
Currently accessTokens are set with 1min expiration, and refreshTokens with 2min expiration for testing.

Also added endpoints to verify validity of accessToken, refreshToken, and regenerate accessToken.
- Added redis cache folder with its Dockerfile, modified to run with docker compose.
Uses redis cache for whitelisted refreshTokens.

- **Note**: Please modify your `.env` file accordingly to add a new REFRESH_TOKEN_SECRET

### Additional information
1. User logs in, user-service issues a new accessToken and refreshToken stored in cookies.
2. User access other services with accessTokens and is allowed while accessToken is still valid.
3. When accessToken expires, other services will deny access.
4. Frontend needs to call user-service to regenerate a new accessToken.
5. If refreshToken is still valid, user-service issues a newAccessToken.
- If refreshToken is invalid: either refreshToken expired, or was from a previously logged in user that has logged out. User needs to re-login again to get a new refreshToken.
6. User can now access other services again.

image

### CODE REVIEW

File: user-service/middlewares/redis.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { createClient } from 'redis';
+
+const client = createClient({
+    url: 'redis://localhost:6379' // Adjust the URL if using Docker or a remote instance
```
Comment (georgetayqy) [inline user-service/middlewares/redis.js:4]: Should we use an environment variable here to allow us to switch endpoints during deployment?

**Review by georgetayqy — CHANGES_REQUESTED** [review]
Good job on implementing the refresh token feature! There are some nits to resolve before we can merge this PR!

**Review by georgetayqy — APPROVED** [review]
LGTM

**Comment by HollaG** [comment]
Merged in #371f5b82152b1c2659f7759320f28b23afff3382

**Pull request closed without merging**